octopusdeploy_framework/resource_machine_proxy.go

```diff
@@ -0,0 +1,143 @@
+	"github.com/OctopusDeploy/terraform-provider-octopusdeploy/octopusdeploy_framework/util"
+	"github.com/hashicorp/terraform-plugin-framework/resource"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+)
+
+var _ resource.Resource = &machineProxyResource{}
+
+type machineProxyResource struct {
+	*Config
+}
+
+func NewMachineProxyResource() resource.Resource {
+	return &machineProxyResource{}
+}
+
+func (r *machineProxyResource) Metadata(_ context.Context, req resource.MetadataRequest, resp *resource.MetadataResponse) {
+	resp.TypeName = util.GetTypeName(schemas.MachineProxyResourceName)
+}
+
+func (r *machineProxyResource) Schema(_ context.Context, _ resource.SchemaRequest, resp *resource.SchemaResponse) {
+	resp.Schema = schemas.MachineProxySchema{}.GetResourceSchema()
+}
+
+func (r *machineProxyResource) Configure(_ context.Context, req resource.ConfigureRequest, resp *resource.ConfigureResponse) {
+	r.Config = ResourceConfiguration(req, resp)
+}
+
+func (r *machineProxyResource) Create(ctx context.Context, req resource.CreateRequest, resp *resource.CreateResponse) {
+	var plan schemas.MachineProxyResourceModel
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &plan)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	machineProxy := mapMachineProxyModelToRequest(&plan)
+	createdProxy, err := proxies.Add(r.Client, machineProxy)
+	if err != nil {
+		resp.Diagnostics.AddError("Error creating machine proxy", err.Error())
+		return
+	}
+
+	proxyModel := mapMachineProxyRequestToModel(createdProxy, &plan)
+
+	diags := resp.State.Set(ctx, proxyModel)
+	resp.Diagnostics.Append(diags...)
+}
+
+func (r *machineProxyResource) Read(ctx context.Context, req resource.ReadRequest, resp *resource.ReadResponse) {
+	var state schemas.MachineProxyResourceModel
+	diags := req.State.Get(ctx, &state)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	machineProxy, err := proxies.GetByID(r.Client, state.SpaceID.ValueString(), state.ID.ValueString())
+	if err != nil {
+		if err := errors.ProcessApiErrorV2(ctx, resp, state, err, "machine proxy"); err != nil {
+			resp.Diagnostics.AddError("Error reading machine proxy", err.Error())
+		}
+		return
+	}
+
+	proxyModel := mapMachineProxyRequestToModel(machineProxy, &state)
+
+	resp.Diagnostics.Append(resp.State.Set(ctx, proxyModel)...)
+}
+
+func (r *machineProxyResource) Update(ctx context.Context, req resource.UpdateRequest, resp *resource.UpdateResponse) {
+	var plan schemas.MachineProxyResourceModel
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &plan)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	existingProxy, err := proxies.GetByID(r.Client, plan.SpaceID.ValueString(), plan.ID.ValueString())
+	if err != nil {
+		resp.Diagnostics.AddError("Error retrieving machine proxy", err.Error())
+		return
+	}
+
+	updatedProxy := mapMachineProxyModelToRequest(&plan)
+	updatedProxy.ID = existingProxy.ID
+	updatedProxy.Links = existingProxy.Links
```

Comment:
Eww :) do we have to? I assume so, but would love if TFP didn't need to concern itself with this detail.

Reply:
Same. At least its not exposed to user.